### PR TITLE
Add ephemeral  dynamic runner workflow

### PR DESCRIPTION
**Pull request by pavlovic-ivan**
**Intention**

The intention of this PR is to provide the project the ability to deploy dynamic runners that will run CUDA jobs, by adding new runners workflow. In order to have the complete feature, Github repo needs to be configured. More about that bellow in the Guide.

**Guide**

This guide explains how to configure Github repo. First, the prerequisits. You need a protected environment. If none exist:

1. Go to `Settings` > `Environments` > `New environment`
2. Give the Environment a name `cuda-test-infra`
3. Select a branch for which this Environment will be available by selecting the option "Selected branches" in the drop down menu of `Deployment branches`.
4. Select "+ Add deployment branch rule"
5. In the popup window enter the name of the main branch (eg, main, master, whichever is the main for your repo)

Now, make sure you have a Personal Access Token (PAT) set. If not:

1. Open Your Profile > `Settings` > `Developer Settings` > `Personal access tokens`
2. Select `Generate new token`
3. Give the token a name
4. Set expiration to "No expiration"
5. Select scopes for the token: `repo: Full control of private repositories`
6. Select `Generate token`
7. Copy the token and save it, it is visible to you only once

After the Environment [exists/is created], add new Environment secrets (provided safely offline), but:

- Select "+ Add Secret" within selected Environment (Environment named `cuda-test-infra`)

**Working example**

A working example can be seen [here](https://github.com/pavlovic-ivan/ILGPU-fork2/actions/runs/1848580697). You will see that CUDA tests are green. Workflow failed on `publish-preview` job, as i don't have perms and secrets needed to make it successful.

When this workflow was created, it created an event that triggered a workflow that deploys the runners. Workflow available [here](https://github.com/pavlovic-ivan/ILGPU-fork2/actions/runs/1848580754). Once the main CI workflow is done, another event triggers a workflow that destroys the created runners (as seen [here](https://github.com/pavlovic-ivan/ILGPU-fork2/actions/runs/1848719253)).

**Further steps before accepting PR**

- [ ] Communicate securely with upstream maintainer. Provide secrets needed to deploy runners to GR GCP project

### CODE REVIEW

**Comment by ljubon** [comment]
Nicely done Ivane! Learned a few neat stuff from your PR.. thank you! 👍 🥇

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -102,7 +118,7 @@ jobs:
   build:
     # Do not run this job for pull requests where both branches are from the same repo.
     # Jobs that depend on this one will be skipped too.
-    # This prevents duplicate CI runs for our own pull requests, whilst preserving the ability to
+    # This prevents duplicate CI runs for our own pull requests, whilst preserving the ability to
```
Comment (jgiannuzzi) [inline .github/workflows/ci.yaml:121]:
```suggestion
    # This prevents duplicate CI runs for our own pull requests, whilst preserving the ability to
```

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -159,8 +175,8 @@ jobs:
         shell: bash
         run: echo ::set-output name=flavor::$([[ "${{ matrix.os }}" == "cuda" ]] && echo "Cuda" || echo "CPU")
 
-      - name: Set values
-        id: values
+      - name: Prepare environment
+        id: prepare-env
         shell: bash
         run: |
           echo "::set-output name=json::{\"ILGPU_CLEAN_TESTS\":\"false\", \"ILGPU_TEST_PARALLELIZATION_LEVEL\":\"1\"}"
```
Comment (jgiannuzzi) [inline .github/workflows/ci.yaml:182]: This should not be part of this PR, as it should be covered by https://github.com/m4rs-mt/ILGPU/pull/712
@ljubon could you please update the PR so it can be reviewed and merged? (add description, change job name and id, set empty json as default, echo only once)

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -433,4 +449,4 @@ jobs:
           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
 
       - name: Publish to NuGet
-        run: dotnet nuget push "*.nupkg" --api-key ${{ secrets.NUGET_API_KEY }} --source https://api.nuget.org/v3/index.json
+        run: dotnet nuget push "*.nupkg" --api-key ${{ secrets.NUGET_API_KEY }} --source https://api.nuget.org/v3/index.json
```
Comment (jgiannuzzi) [inline .github/workflows/ci.yaml:452]: Make sure you don't change the existing formatting

File: .github/workflows/runners.yaml [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    runs-on: ubuntu-latest
+    steps:
+      - name: Detect fork status
+        id: detect-fork-status
+        run: |
+          upstream=false
+          
+          (
+            [ "${{ github.event.repository.fork }}" == "false" ] || 
+            [ $GITHUB_REPOSITORY == "${{ github.repository }}" ]
```
Comment (jgiannuzzi) [inline .github/workflows/runners.yaml:30]: I _think_ this will always evaluate to `true`. Is there a reason why you don't use the same logic as in `ci.yml` to check whether it's a fork?

Comment (jgiannuzzi) [reply .github/workflows/runners.yaml:30]: Also worth checking whether this is needed at all — maybe `github.event.repository.fork` is enough for `workflow_run` events?

File: .github/workflows/runners.yaml [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        uses: actions/checkout@v2
+        with:
+          repository: pavlovic-ivan/ephemeral-github-runner-gcp
+          ref: main
+
+      - name: Download config
+        uses: actions/download-artifact@v2
+        with:
+          name: pulumi-config
+          path: artifacts
```
Comment (jgiannuzzi) [inline .github/workflows/runners.yaml:68]: Instead of uploading and downloading the config, you could checkout the 2 repos in different directories, and then copy the config over

File: .github/workflows/runners.yaml [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+      - completed
+      - requested
+
+env:
+  PAT: ${{ secrets.PAT }}
+  GOOGLE_PROJECT: ${{ secrets.GOOGLE_PROJECT }}
+  PULUMI_CONFIG_PASSPHRASE: ${{ secrets.PULUMI_CONFIG_PASSPHRASE }}
+  GOOGLE_CREDENTIALS: ${{ secrets.GCE_SA_KEY }}
+  GOOGLE_REGION: ${{ secrets.GOOGLE_REGION }}
+  PULUMI_STACK_NAME: ${{ github.event.workflow_run.id }}
```
Comment (jgiannuzzi) [inline .github/workflows/runners.yaml:16]: maybe we could prefix this with the repo name as well?

File: .github/workflows/runners.yaml [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        uses: actions/checkout@v2
+        with:
+          repository: pavlovic-ivan/ephemeral-github-runner-gcp
+          ref: main
+
+      - name: Download config
+        uses: actions/download-artifact@v2
+        with:
+          name: pulumi-config
+          path: artifacts
```
Comment (jgiannuzzi) [inline .github/workflows/runners.yaml:114]: Same remark as above

File: config.yaml [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+  ephemeral-github-runner:bootDiskSizeInGB: 200
+  ephemeral-github-runner:bootDiskType: pd-balanced
+  ephemeral-github-runner:env: dev
+  ephemeral-github-runner:ghRunnerName: orville
+  ephemeral-github-runner:machineImage: projects/gr-oss-251320/global/images/ubuntu-2004-focal-v20211202-ghrv22851
+  ephemeral-github-runner:machineType: a2-highgpu-1g
+  ephemeral-github-runner:repo: ILGPU
+  ephemeral-github-runner:repoOwner: m4rs-mt
+  ephemeral-github-runner:runnersCount: 4
+  ephemeral-github-runner:serviceAccountEmail: pulumi@gr-oss-251320.iam.gserviceaccount.com
```
Comment (jgiannuzzi) [inline config.yaml:11]: [optional] could this be made a secret?

File: .github/workflows/ci.yaml [hunk 9]
```diff
@@ -6,12 +6,27 @@ on:
...
+        id: check-if-upstream
+        run: |
+          fork=$(gh api repos/${{ github.repository }} | jq .fork)
+          upstream=false
+          [ $fork == "false" ] && upstream=true
+          echo "::set-output name=upstream::$upstream"
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+    outputs:
+      upstream: ${{ steps.check-if-upstream.outputs.upstream }}
```
[the remark this answers is not in the document]
Comment (pavlovic-ivan) [reply .github/workflows/ci.yaml:29]: I get what you mean. This is where my developer side kicked in. I don't like negation. Renaming this to `is-fork` means that double negation will be used as well down the line. To me, `is-fork` is equal to `is not upstream`. To check if it is upstream, i would have to do `not(is-fork)`, translating into `is not not upstream`. You know what i mean? 😄

File: .github/workflows/ci.yaml [hunk 10]
```diff
@@ -6,12 +6,27 @@ on:
   schedule:
     # Run daily at 00:00 so we get notified if CI is broken before a pull request
     # is submitted.
-    - cron:  '0 0 * * *'
+  - cron: '0 0 * * *'
```
[the remark this answers is not in the document]
Comment (pavlovic-ivan) [reply .github/workflows/ci.yaml:9]: Understood, thanks!

File: .github/workflows/ci.yaml (hunk 2 above)
Comment (pavlovic-ivan) [reply .github/workflows/ci.yaml:182]: All right. Then, the #712 should be merged first, before this one. I will need to wait for that, no? I see it is currently a draft. @ljubon please add this PR to Ora board as well if it's not there

File: .github/workflows/ci.yaml (hunk 3 above)
Comment (pavlovic-ivan) [reply .github/workflows/ci.yaml:445]: Not sure if new line is the problem or not. I actually didn't change here anything. Could be that different new line characters were introduced automatically by my OS (if it is different than Marcel's)

File: .github/workflows/ci.yaml [hunk 11]
```diff
@@ -72,6 +87,7 @@ jobs:
     # run the CI for each branch push to a fork, and for each pull request originating from a fork.
     if: github.event_name == 'schedule' || github.event_name == 'push' || github.event.pull_request.head.repo.id != github.event.pull_request.base.repo.id
     runs-on: ubuntu-latest
+    needs: is-upstream
```
[the remark this answers is not in the document]
Comment (pavlovic-ivan) [reply .github/workflows/ci.yaml]: Agreed

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (pavlovic-ivan) [reply .github/workflows/ci.yaml]: Should be resolved now

File: .github/workflows/runners.yaml (hunk 4 above)
Comment (pavlovic-ivan) [reply .github/workflows/runners.yaml:30]: I'll check, thanks

File: .github/workflows/runners.yaml (hunk 5 above)
Comment (pavlovic-ivan) [reply .github/workflows/runners.yaml:68]: Sounds good

File: .github/workflows/runners.yaml (hunk 6 above)
Comment (pavlovic-ivan) [reply .github/workflows/runners.yaml:16]: Yes, we can. I'll add that as well

File: .github/workflows/runners.yaml (hunk 7 above)
Comment (pavlovic-ivan) [reply .github/workflows/runners.yaml:114]: All right, thanks :)

File: config.yaml (hunk 8 above)
Comment (pavlovic-ivan) [reply config.yaml:11]: Yes, of course. I'll make it a secret

File: .github/workflows/ci.yaml [hunk 12]
```diff
@@ -137,7 +146,7 @@ jobs:
           - os: windows-latest
             library: ILGPU.Algorithms
             framework: net471
-      fail-fast: true
+      fail-fast: false
```
Comment (ljubon) [inline .github/workflows/ci.yaml:149]: is there any specific reason why this is set to `false`?

File: .github/workflows/runners.yaml (hunk 5 above)
Comment (ljubon) [reply .github/workflows/runners.yaml:68]: FYI same was done for chart repo [here](https://github.com/ljubon/charts/blob/refactor-gh-pages/.github/workflows/publish-charts.yaml#L46) 👀

Comment (pavlovic-ivan) [reply .github/workflows/runners.yaml:68]: @ljubon thanks a bunch! 🙇‍♂️

File: .github/workflows/ci.yaml (hunk 12 above)
Comment (pavlovic-ivan) [reply .github/workflows/ci.yaml:149]: This was an accident. I'll revert. Thanks for the sharp eye!

File: .github/workflows/ci.yaml (hunk 2 above)
Comment (ljubon) [reply .github/workflows/ci.yaml:175]: PR updated https://github.com/m4rs-mt/ILGPU/pull/712 🤞

File: .github/workflows/ci.yaml (hunk 9 above)
[the remark this answers is not in the document]
Comment (jgiannuzzi) [reply .github/workflows/ci.yaml]: I get your point, but here you are essentially implementing `is-upstream` as `not(is-fork)`, just so you can do `not(not(is-fork))` further. The 4 lines of shell would be just one if this was `is-fork`:

```yaml
run: echo "::set-output name=fork::$(gh api repos/${{ github.repository }} | jq .fork)"
```

File: .github/workflows/ci.yaml (hunk 3 above)
Comment (jgiannuzzi) [reply .github/workflows/ci.yaml:445]: It's your editor that is acting differently from Marcel's. It's best to avoid those changes, as they detract from the actual content of the pull request. It would also create a flip-flop of formatting between contributors.
One way to solve this is to rewrite your commit with [GitUp](https://gitup.co) and discard this line change.

File: .github/workflows/ci.yaml (hunk 2 above)
Comment (jgiannuzzi) [reply .github/workflows/ci.yaml:175]: @pavlovic-ivan I see that the `master` branch of your fork is not in line with the upstream repo, but rather has an old version of @ljubon's branch for https://github.com/m4rs-mt/ILGPU/pull/712.

Could you please either update your `master` branch, or reset it to the latest upstream `master`?

File: .github/workflows/runners.yaml [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+
+env:
+  PAT: ${{ secrets.PAT }}
+  GOOGLE_PROJECT: ${{ secrets.GOOGLE_PROJECT }}
+  PULUMI_CONFIG_PASSPHRASE: ${{ secrets.PULUMI_CONFIG_PASSPHRASE }}
+  GOOGLE_CREDENTIALS: ${{ secrets.GCE_SA_KEY }}
+  GOOGLE_REGION: ${{ secrets.GOOGLE_REGION }}
+  PULUMI_STACK_NAME: ilgpu_${{ github.event.workflow_run.id }}
+  GS_BUCKET: ${{ secrets.GS_BUCKET }}
+  GCE_SA_EMAIL: ${{ secrets.GCE_SA_EMAIL }}
```
Comment (jgiannuzzi) [inline .github/workflows/runners.yaml:18]: I would avoid exposing all those secrets to the steps that don't need it. We could maybe move them to the `Execute pulumi` steps?

File: .github/workflows/runners.yaml [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+  is-fork:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check if repo is fork
+        id: is-fork
+        run: echo "::set-output name=fork::$(gh api repos/${{ github.repository }} | jq .fork)"
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+    outputs:
+      fork: ${{ steps.is-fork.outputs.fork }}
```
Comment (jgiannuzzi) [inline .github/workflows/runners.yaml:30]: Could you please check whether this is necessary or whether the information is already somewhere under `${{ github }}`? Maybe it will still be necessary for `workflow_run` events triggered by `scheduled` workflow runs, in which case, there would be nothing to change!

File: .github/workflows/runners.yaml [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+      - name: Checkout
+        uses: actions/checkout@v2
+        with:
+          path: ilgpu
+
+      - name: Checkout Pulumi Ephemeral Github Runner project
+        uses: actions/checkout@v2
+        with:
+          repository: pavlovic-ivan/ephemeral-github-runner-gcp
+          ref: main
```
Comment (jgiannuzzi) [inline .github/workflows/runners.yaml:46]: note for future improvement (_not_ for now): this should probably be a tag instead, so we use a stable version of the repo

File: .github/workflows/runners.yaml [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+          node-version: '14'
+
+      - name: Execute Pulumi
+        run: |
+          cd pulumi
+          curl -fsSL https://get.pulumi.com | sh
+          npm install
+          pulumi login $GS_BUCKET
+          pulumi stack init -s $PULUMI_STACK_NAME
+          pulumi up --diff --config-file ../ilgpu/config.yaml -y
```
Comment (jgiannuzzi) [inline .github/workflows/runners.yaml:61]: I would like this step to be split between installing pulumi, logging into pulumi, and then the actual deployment with pulumi

File: .github/workflows/runners.yaml [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+
+      - name: Execute Pulumi
+        run: |
+          cd pulumi
+          curl -fsSL https://get.pulumi.com | sh
+          npm install
+          pulumi login $GS_BUCKET
+          pulumi stack select $PULUMI_STACK_NAME
+          pulumi destroy --config-file artifacts/config.yaml -y
+          pulumi stack rm $PULUMI_STACK_NAME -y
```
Comment (jgiannuzzi) [inline .github/workflows/runners.yaml:89]: These are the only different lines between this job and the `deploy-runners` job. I think we should make it all one job, and then have a different behaviour in the 'Deploy with Pulumi' step based on the action (`requested` or `completed`).

File: config.yaml [hunk 18]
```diff
@@ -0,0 +1,12 @@
+config:
+  ephemeral-github-runner:bootDiskSizeInGB: 200
+  ephemeral-github-runner:bootDiskType: pd-balanced
+  ephemeral-github-runner:env: dev
+  ephemeral-github-runner:ghRunnerName: orville
+  ephemeral-github-runner:machineImage: projects/gr-oss-251320/global/images/ubuntu-2004-focal-v20211202-ghrv22851
+  ephemeral-github-runner:machineType: a2-highgpu-1g
+  ephemeral-github-runner:repo: ILGPU
+  ephemeral-github-runner:repoOwner: m4rs-mt
```
Comment (jgiannuzzi) [inline config.yaml:9]: nitpick: can those maybe be deduced at runtime?

File: .github/workflows/runners.yaml [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+          path: pulumi
+
+      - name: Execute Pulumi
+        run: |
+          cd pulumi
+          curl -fsSL https://get.pulumi.com | sh
+          npm install
+          pulumi login $GS_BUCKET
+          pulumi stack select $PULUMI_STACK_NAME
+          pulumi destroy --config-file artifacts/config.yaml -y
```
Comment (jgiannuzzi) [inline .github/workflows/runners.yaml:88]: This is the wrong path the config file
```suggestion
          pulumi destroy --config-file ../ilgpu/config.yaml -y
```

File: .github/workflows/runners.yaml (hunk 14 above)
Comment (jgiannuzzi) [reply .github/workflows/runners.yaml:30]: See https://github.com/pavlovic-ivan/ILGPU/pull/4#discussion_r783767462 as well

File: .github/workflows/runners.yaml (hunk 13 above)
Comment (pavlovic-ivan) [reply .github/workflows/runners.yaml:18]: Avreed

File: .github/workflows/runners.yaml (hunk 14 above)
Comment (pavlovic-ivan) [reply .github/workflows/runners.yaml:30]: Will do, but i think this was already investigated, and the only reason why this is the implementation, is exactly as you've said, because of `workflow_run` events

File: .github/workflows/runners.yaml (hunk 15 above)
Comment (pavlovic-ivan) [reply .github/workflows/runners.yaml:46]: yes, correct, currently there are no tags in the mentioned repo

File: .github/workflows/runners.yaml (hunk 16 above)
Comment (pavlovic-ivan) [reply .github/workflows/runners.yaml:61]: sounds good!

File: .github/workflows/runners.yaml (hunk 19 above)
Comment (pavlovic-ivan) [reply .github/workflows/runners.yaml:88]: Thanks for pointing this out, i was sure i've updated the path to the file!

File: config.yaml (hunk 18 above)
Comment (pavlovic-ivan) [reply config.yaml:9]: I am sure it can, however i would have to investigate. Either those vars can get overriden inline when executing Pulumi, or i will need to pass this via ENV vars

File: .github/workflows/runners.yaml (hunk 14 above)
Comment (pavlovic-ivan) [reply .github/workflows/runners.yaml:20]: Ok, so i've checked the github context and here is the situations:

Upstream Repo - on manual push:
- main workflow -> `github.event.repository.fork = false`
- on workflow_run -> `github.event.repository.fork = false`

Upstream Repo - on scheduled run:
- main workflow -> `github.event.repository.fork = /` doesn't exist. github event object looks like this 
`github: {
  event: {
    schedule: */15 * * * *
  }
}`
- on workflow_run -> `github.event.repository.fork = false`

Forked repo - on manual push:
- main workflow -> `github.event.repository.fork = true`
- on workflow_run -> `github.event.repository.fork = true`

Forked repo - on scheduled run:
- main workflow -> `github.event.repository.fork = /` doesn't exist. github event object looks like this 
`github: {
  event: {
    schedule: */15 * * * *
  }
}`
- on workflow_run -> `github.event.repository.fork = true`

There is a bug apparently in the github.event context only for the main workflow. Event object is missing all kinds of data, it only contains the `schedule` fields set to what ever is set in the yaml file. I remember this issue when was adding cuda jobs to the scheduled runs. For the time being, i think this particular line of code will need to stay.

File: .github/workflows/runners.yaml (hunk 15 above)
Comment (pavlovic-ivan) [reply .github/workflows/runners.yaml:36]: @jgiannuzzi do you feel like we can resolve this conversation

**Comment by pavlovic-ivan** [comment]
@jgiannuzzi you will notice that the last step in the `manage-runners` job, that executes pulumi, is actually now running a `make` command. I've added Makefile to ephemeral runners pulumi repo, to act as a wrapper to all those nasty lines of pulumi commands. Hope you don't mind! Looking forward for the feedback. I think now all tasks are resolved 🙏

File: .github/workflows/runners.yaml (hunk 15 above)
Comment (jgiannuzzi) [reply .github/workflows/runners.yaml:36]: sure!

File: .github/workflows/runners.yaml (hunk 14 above)
Comment (jgiannuzzi) [reply .github/workflows/runners.yaml:20]: as `github.event.repository.fork` always has the right value for `workflow_run`, I would say we should get rid of this whole extra job

File: .github/workflows/runners.yaml [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+        working-directory: pulumi
+        run: |
+          if [[ ${{ github.event.action }} == 'requested' ]]; then
+            make up stack=$PULUMI_STACK_NAME config=../ilgpu/config.yaml auto-approve=-y
+          elif [[ ${{ github.event.action }} == 'completed' ]]; then
+            make down stack=$PULUMI_STACK_NAME config=../ilgpu/config.yaml auto-approve=-y
+          else
+            echo "Workflow failed. Unrecognized workflow event received: github.event.action=${{ github.event.action }}"
+            exit 1
+          fi
```
Comment (jgiannuzzi) [inline .github/workflows/runners.yaml:76]: I think that this would be better expressed with a `case` which just sets the make target (`up` or `down`), followed by only one `make $target stack=...` command.
I also wonder whether we should actually treat the case of the unknown workflow event, as we explicitly filter them at the top of the file. If you really want to treat that case, you should at least make the error a proper workflow error (see the [docs](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-error-message)).

File: config.yaml [hunk 21]
```diff
@@ -0,0 +1,10 @@
+config:
+  ephemeral-github-runner:bootDiskSizeInGB: 200
+  ephemeral-github-runner:bootDiskType: pd-balanced
+  ephemeral-github-runner:env: dev
```
Comment (jgiannuzzi) [inline config.yaml:4]: I'm not sure what this does — is this required? should it be `dev`?

File: config.yaml [hunk 22]
```diff
@@ -0,0 +1,10 @@
...
+  ephemeral-github-runner:ghRunnerName: orville
```
Comment (jgiannuzzi) [inline config.yaml:5]: I'm not sure what this does either — it this required? why is it `orville`?

File: config.yaml [hunk 23]
```diff
@@ -0,0 +1,10 @@
...
+  ephemeral-github-runner:machineImage: projects/gr-oss-251320/global/images/ubuntu-2004-focal-v20211202-ghrv22851
+  ephemeral-github-runner:machineType: a2-highgpu-1g
+  ephemeral-github-runner:runnersCount: 6
+  ephemeral-github-runner:serviceAccountScopes: https://www.googleapis.com/auth/cloud-platform
+  ephemeral-github-runner:zone: europe-west4-a
```
Comment (jgiannuzzi) [inline config.yaml:10]: as we set the region as a secret, should we have the zone? should we even specify it? should we not let GCP distribute the VM in the zone of its choice instead?

**Review by jgiannuzzi — COMMENTED** [review]
Great improvements, well done @pavlovic-ivan!

File: config.yaml (hunk 22 above)
Comment (pavlovic-ivan) [reply config.yaml:5]: It is required as it will be in the instance name visible in GCP. I was trying to be a bit creative about the name, there's a sci fi show called Orville, so that's why :) If you think it's best to rename it to `ghrunner`, i can do that as well

File: config.yaml (hunk 23 above)
Comment (pavlovic-ivan) [reply config.yaml:10]: I'll test that out, but if i remember correctly, it is necessary even though we have a region set

File: .github/workflows/runners.yaml (hunk 20 above)
Comment (pavlovic-ivan) [reply .github/workflows/runners.yaml:76]: Gotcha. Ok, thanks, i will refactor it like that. For the error case, i will remove it entirely, as you've said, with workflow events it should be filtered out already

File: .github/workflows/runners.yaml (hunk 14 above)
Comment (pavlovic-ivan) [reply .github/workflows/runners.yaml:20]: Ah, yes. You are right. For this workflow the entire job is unnecessary. I didn't see that, even though i've written it down...

File: config.yaml (hunk 23 above)
Comment (pavlovic-ivan) [reply config.yaml:10]: @jgiannuzzi do you see any action points here or can we close this discussion?

File: config.yaml (hunk 22 above)
Comment (jgiannuzzi) [reply config.yaml]: understood — I think the new name you came up with is more descriptive 👍

File: config.yaml (hunk 23 above)
Comment (jgiannuzzi) [reply config.yaml:9]: could this be a GCP thing, e.g. is there maybe a special 'automatic' zone?

Comment (pavlovic-ivan) [reply config.yaml:9]: @jgiannuzzi  Honestly, i am not sure. Maybe this can be a part of a future improvement? In the meantime, we investigate if this is possible. How do you feel about that?

Comment (pavlovic-ivan) [reply config.yaml:9]: I'll add it as a secret. I feel that would be best :)

**Comment by pavlovic-ivan** [comment]
🎉 🎉 🎉 

Awesome! Let's make some people happy :)

**Comment by pavlovic-ivan** [comment]
@jgiannuzzi @ljubon please also take a look at the top-most comment i've made, with instructions. I've updated it just now. I've removed the list of secrets that need to be set, as they will be provided offline securely. 🙏

**Review by ljubon — APPROVED** [review]
Awesome! 💯
